packages/upload-api/src/blob/add.js

```diff
@@ -0,0 +1,216 @@
+import * as W3sBlob from '@web3-storage/capabilities/web3.storage/blob'
+import * as HTTP from '@web3-storage/capabilities/http'
+import * as UCAN from '@web3-storage/capabilities/ucan'
+import * as API from '../types.js'
+
+import { BlobExceedsSizeLimit } from './lib.js'
+
+/**
+ * @param {API.BlobServiceContext} context
+ * @returns {API.ServiceMethod<API.BlobAdd, API.BlobAddSuccess, API.BlobAddFailure>}
+ */
+export function blobAddProvider(context) {
+  return Server.provideAdvanced({
+    capability: Blob.add,
+    handler: async ({ capability, invocation }) => {
+      const {
+        id,
+        allocationsStorage,
+        maxUploadSize,
+        getServiceConnection,
+        tasksStorage,
+      } = context
+      const { blob } = capability.nb
+      const space = /** @type {import('@ucanto/interface').DIDKey} */ (
+        Server.DID.parse(capability.with).did()
+      )
+
+      // Verify blob is within accept size
+      if (blob.size > maxUploadSize) {
+        return {
+          error: new BlobExceedsSizeLimit(maxUploadSize),
+        }
+      }
+
+      // We derive principal from the content multihash to be an audience
+      // of the `http/put` invocation. That way anyone with blob content
+      // could perform the invocation and issue receipt by deriving same
+      // principal
+      const putSubject = await ed25519.derive(blob.content.slice(0, 32))
+      const facts = [
+        {
+          keys: putSubject.toArchive(),
+        },
+      ]
+
+      // Create web3.storage/blob/* invocations
+      const blobAllocate = W3sBlob.allocate.invoke({
+        issuer: id,
+        audience: id,
+        with: id.did(),
+        nb: {
+          blob,
+          cause: invocation.link(),
+          space,
+        },
+        expiration: Infinity,
+      })
+      const blobAccept = W3sBlob.accept.invoke({
+        issuer: id,
+        audience: id,
+        with: id.toDIDKey(),
+        nb: {
+          blob,
+          exp: Number.MAX_SAFE_INTEGER,
+        },
+        expiration: Infinity,
+      })
+      const [allocatefx, acceptfx] = await Promise.all([
+        blobAllocate.delegate(),
+        blobAccept.delegate(),
+      ])
+
+      // Get receipt for `blob/allocate` if available, or schedule invocation if not
+      const allocatedGetRes = await allocationsStorage.get(space, blob.content)
+      let blobAllocateReceipt
+      let blobAllocateOutAddress
+      // If already allocated, just get the allocate receipt
+      // and the addresses if still pending to receive blob
+      if (allocatedGetRes.ok) {
+        const receiptGet = await context.receiptsStorage.get(allocatefx.link())
+        if (receiptGet.error) {
+          return receiptGet
+        }
+        blobAllocateReceipt = receiptGet.ok
+
+        // Check if despite allocated, the blob is still not stored
+        const blobHasRes = await context.blobsStorage.has(blob.content)
+        if (blobHasRes.error) {
+          return blobHasRes
```

Comment:
@Gozala any suggestion on how to track if expired? 

in the URL it is encoded a `X-Amz-Date=20240409T155511Z&X-Amz-Expires=86400`, but I would like to avoid requiring this to be tied to a provider. Also, I would like to try to only create presigned URL if it allocates, which also makes it difficult to just store the state in allocation table. I wonder then, if it could be an option to stick it in the receipt, but also annoying to not be tied to the presigned url encoding. Also AWS presigned URL do not allow us to specify an expiration date, but the duration after presigned URL is created.

What do you think?